### PR TITLE
add kvstore balance greed alloc straregy

### PR DESCRIPTION
**Pull request by junranhe**
kvstore_device上更好的均匀分配在gpu上，在gpu显存小，网络大的时候比较明显，在3个3G 780, 36 batch_size 时，vggnet11layer最大占用2800m降低低到2660m，inception-bn 96batch显存差距缩小到10m内

### CODE REVIEW

File: python/mxnet/model.py [hunk 1]
```diff
@@ -261,6 +261,24 @@ def kv_updater(index, grad, weight):
     merged_shape = tuple(merged_shape)
     out_cpu_array = nd.zeros(merged_shape, cpu())
 
+    grad_update_order = []
+    for index, pair in enumerate(zip(arg_blocks, grad_blocks)):
+        arg_list, grad_list = pair
+        if grad_list[0] is not None:
+            grad_size = 1
+            for d in grad_list[0].shape:
```
Comment (tqchen) [inline python/mxnet/model.py]: Is it possible not to introduce the gradient update order on this side?

We can make use of the space information on C++ side, More comment on PR

**Comment by tqchen** [comment]
Let us only change the kvstore_device

If we want to get a global information. override the Init Function https://github.com/dmlc/mxnet/blob/master/src/kvstore/kvstore_local.h#L31 

```c++
void Init(const std::vector<int>& keys,
             const std::vector<NDArray>& values) override {
    // call kvstore local init 
    KVStoreLocal::Init(keys, values);
   // add the strategy here. for example, assign each key to context. You can get size from values[i].shape().Size()
}
```

**Comment by tqchen** [comment]
Also please check https://travis-ci.org/dmlc/mxnet/builds/84667681  The python error seems can be removed by only changing kvstore_device. The lint style check can be reproduced locally by ```make lint```

https://mxnet-bing.readthedocs.org/en/latest/contribute.html

**Comment by junranhe** [comment]
好的，我知道改Init是最优雅的实现，不过要改一下Init的接口，把contexts传进来

**Comment by tqchen** [comment]
Let me see, do you really need context? 

Since the update function is in a first come first allocate strategy. You can compute the information you need (Size of the variables in Init), and do allocation in Merge

**Comment by tqchen** [comment]
It would be great if everything can be done in current interface. I guess it is possible

**Comment by tqchen** [comment]
@junranhe Let us now if you have any updates. Hopefully we can do it by only changing implement the function Init and Merge

**Comment by junranhe** [comment]
kvstore_device 改为init的时候记录下所有values的shape，第一次update的时候初始化所有mergebuf，ps：代码里是不是还有其他地方会影响每个设备分配内存的大小的？发现每次起来极其，还是显存分布还是有些浮动

File: src/kvstore/kvstore_device.h [hunk 2]
```diff
@@ -22,22 +22,56 @@ namespace kvstore {
...
  protected:
+  using KeyShape = std::pair<int, TShape>;
+  void Init(const std::vector<int>& keys,
+            const std::vector<NDArray>& values) override {
+    KVStoreLocal::Init(keys, values);
+
+    for (size_t i = 0; i < keys.size(); ++i) {
+      sorted_key_shape_.push_back(std::make_pair(keys[i], values[i].shape()));
+    }
+    std::sort(sorted_key_shape_.begin(), sorted_key_shape_.end(), [](
```
Comment (tqchen) [inline src/kvstore/kvstore_device.h]: move the sort to allocation function

File: src/kvstore/kvstore_device.h [hunk 3]
```diff
@@ -22,22 +22,56 @@ namespace kvstore {
...
   const NDArray& MergePushValue(
       int key, const std::vector<NDArray>& val, int priority) override {
     if (updater_ != nullptr) {
       // fall back to CPU based update if updater presents
       return KVStoreLocal::MergePushValue(key, val, priority);
     }
+
+    if (merge_buf_.empty()) {
+      CHECK(!val.empty());
+      std::unordered_map<int, std::pair<Context, size_t>> ctx_info;
```
Comment (tqchen) [inline src/kvstore/kvstore_device.h]: move sorted here

File: src/kvstore/kvstore_device.h (hunk 2 above)
Comment (tqchen) [inline src/kvstore/kvstore_device.h]: Init can be called multiple times

File: src/kvstore/kvstore_device.h [hunk 4]
```diff
@@ -22,22 +22,56 @@ namespace kvstore {
...
-    if (buf.merged.is_none()) {
-      // round robin to assign the reduction device.
-      size_t round_robin_index =
-          std::uniform_int_distribution<size_t>(0, val.size()-1)(rnd_);
-      buf.ctx = val[round_robin_index].ctx();
-      buf.merged = NDArray(val[0].shape(), buf.ctx);
-    }
-    CopyFromTo(val[0], &(buf.merged), priority);
+    CHECK(!buf.merged.is_none());
+        CopyFromTo(val[0], &(buf.merged), priority);
```
Comment (tqchen) [inline src/kvstore/kvstore_device.h]: remove tab

File: src/kvstore/kvstore_device.h [hunk 5]
```diff
@@ -22,22 +22,56 @@ namespace kvstore {
...
+      std::unordered_map<int, std::pair<Context, size_t>> ctx_info;
+      for (size_t i = 0; i < val.size(); ++i) {
+        int32_t dev_id = val[i].ctx().dev_id;
+        ctx_info[dev_id] = std::make_pair(val[i].ctx(), 0);
+      }
+      for (size_t i = 0; i < sorted_key_shape_.size(); ++i) {
+        int k = sorted_key_shape_[i].first;
+        TShape s = sorted_key_shape_[i].second;
+        auto& tm_buf = merge_buf_[k];
+        size_t min_size = size_t(-1);
```
Comment (tqchen) [inline src/kvstore/kvstore_device.h]: std::numeric_limits<size_t>::max()

**Comment by tqchen** [comment]
Thanks, I have a few comments here. Please also consider move the allocation algorithm to an independent function, so it can be changed easily.

Other than that it looks good to me

**Comment by tqchen** [comment]
One minor thing about current implementation, is that it assumes the Push must be called after all Init calls. This is true for current use-case. It will be nice if the implementation can be improved, to the point such that it could also work for interleaving Init and Push(which is rare).

Thanks for contribution,  I have added your name to https://github.com/dmlc/mxnet/blob/master/CONTRIBUTORS.md  Please feel free to open PR if you like to add more details about you or contribute more to the project

**Comment by junranhe** [comment]
理解，如果交替 init 和 push， 那算法只能是在线的，觉得只有把大权重拆分到多gpu上，再贪心均衡，感觉用ndarray的slice可以比较方便实现部分同步，我这几天工作有点忙，要晚些才能提交一个这样的实现:)

**Pull request merged**